third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,244 @@
+  // available for scratch buffer.
+  int64_t
+  computeTargetScratchBufferSize(triton::gpu::ConvertLayoutOp op,
+                                 Allocation *allocation,
+                                 ArrayRef<Allocation::BufferId> liveBuffers) {
+    int totalSize = 0;
+    auto scratchBufferId = allocation->getBufferId(op.getOperation());
+    int64_t scratchBufferSize = allocation->getAllocatedSize(scratchBufferId);
+    size_t totalLDSConsumption = 0;
+    for (auto buf : liveBuffers) {
+      totalLDSConsumption = std::max(
+          totalLDSConsumption, allocation->getAllocatedInterval(buf).end());
+    }
+    int64_t freeRequired = totalLDSConsumption - LDSLimit;
+    return std::max(static_cast<int64_t>(0), scratchBufferSize - freeRequired);
+  }
+
+  SmallVector<LDSBottleneckOperation>
+  findLDSBottleneckLayoutConvert(ModuleAllocation &allocAnalysis,
+                                 FunctionOpInterface func) {
+    SmallVector<LDSBottleneckOperation> candidates;
+    auto funcAnalysis = allocAnalysis.getFuncData(func);
+    auto liveBuffers = funcAnalysis->getLiveBuffers();
+
+    func.walk([&](triton::gpu::ConvertLayoutOp cvtOp) -> void {
+      auto cvtBuffer = funcAnalysis->getBufferId(cvtOp.getOperation());
+      assert(cvtBuffer != Allocation::InvalidBufferId);
+
+      auto targetScratchBufferSize = computeTargetScratchBufferSize(
+          cvtOp, funcAnalysis, liveBuffers[cvtOp]);
+      auto currentLDSConsumption = funcAnalysis->getAllocatedSize(cvtBuffer);
+      if (currentLDSConsumption > targetScratchBufferSize)
+        candidates.push_back({cvtOp, targetScratchBufferSize});
+    });
+    return candidates;
+  }
+
+public:
+  OptimizeAMDLDSUsage(StringRef targetArch, int customLDSLimit)
+      : OptimizeAMDLDSUsageBase<OptimizeAMDLDSUsage>() {
+    this->targetArch = targetArch.str();
+    this->customLDSLimit = customLDSLimit;
+  }
+
+  void runOnOperation() override {
+    ModuleOp mod = getOperation();
+
+    if ((this->LDSLimit = this->customLDSLimit) == 0) {
+      if (this->targetArch.empty()) {
+        mod->emitError("missing gfx* target for pass ")
+            << this->getName().str();
+        return signalPassFailure();
+      }
+      triton::AMD::TargetInfo targetInfo(this->targetArch.c_str());
+      LDSLimit = targetInfo.getSharedMemorySize();
+    }
+
+    ModuleAllocation allocAnalysis(mod);
+    if (allocAnalysis.getSharedMemorySize() <= LDSLimit)
+      return;
+
+    auto rootFunctions = allocAnalysis.getRoots();
+    for (auto rootFunc : rootFunctions) {
+      // Find operations with peak LDS consumption
+      auto candidates = findLDSBottleneckLayoutConvert(allocAnalysis, rootFunc);
+      // Try to transform candidate operations to fit them into LDS
+      for (auto candidate : candidates)
+        tryFitCvtIntoLDS(candidate.op, candidate.LDSSizeTarget);
+    }
+  }
+};
+
+} // namespace
+
+namespace mlir::triton::AMD {
```

Comment:
@antiagainst 
You changed this part from multiple line to one, is it preferred code style now?